src/ripple/app/misc/impl/ValidatorList.cpp

```diff
@@ -569,154 +569,102 @@ ValidatorList::for_each_listed (
+            << "Using unsafe quorum of "
+            << quorum
+            << " as specified in the command line";
     }
 
-    // This minimum quorum guarantees safe overlap with the trusted sets of
-    // other nodes using the same set of published lists.
-    std::size_t quorum = calculateMinimumQuorum (keyListings_.size(),
-        localPubKey_.size() && !localKeyListed);
-
-    JLOG (j_.debug()) <<
-        rankedKeys.size() << "  of " << keyListings_.size() <<
-        " listed validators eligible for inclusion in the trusted set";
-
-    auto size = rankedKeys.size();
+    return quorum;
+}
 
-    // Require 80% quorum if there are lots of validators.
-    if (rankedKeys.size() > BYZANTINE_THRESHOLD)
-    {
-        // Use all eligible keys if there is only one trusted list
-        if (publisherLists_.size() == 1 ||
-                keyListings_.size() < MINIMUM_RESIZEABLE_UNL)
-        {
-            // Try to raise the quorum to at least 80% of the trusted set
-            quorum = std::max(quorum, size - size / 5);
-        }
-        else
-        {
-            // Reduce the trusted set size so that the quorum represents
-            // at least 80%
-            size = quorum * 1.25;
-        }
-    }
+TrustChanges
+ValidatorList::updateTrusted(hash_set<NodeID> const& seenValidators)
+{
+    boost::unique_lock<boost::shared_mutex> lock{mutex_};
 
-    if (minimumQuorum_ && seenValidators.size() < quorum)
+    // Remove any expired published lists
+    for (auto const& list : publisherLists_)
     {
-        quorum = *minimumQuorum_;
-        JLOG (j_.warn())
-            << "Using unsafe quorum of "
-            << quorum_
-            << " as specified in the command line";
+        if (TimeKeeper::time_point{} < list.second.expiration &&
+            list.second.expiration <= timeKeeper_.now())
+            removePublisherList(list.first);
     }
 
-    // Do not use achievable quorum until lists from all configured
-    // publishers are available
-    else if (! allListsAvailable)
-        quorum = std::numeric_limits<std::size_t>::max();
-
     TrustChanges trustChanges;
     {
         hash_set<PublicKey> newTrustedKeys;
-        for (auto const& val : boost::adaptors::reverse(rankedKeys))
+
+        for (auto const& val : keyListings_)
         {
-            if (size <= newTrustedKeys.size())
-                break;
-            newTrustedKeys.insert(val.second);
+            if (validatorManifests_.revoked (val.first))
+                continue;
 
-            if (trustedKeys_.erase(val.second) == 0)
-                trustChanges.added.insert(calcNodeID(val.second));
+            newTrustedKeys.insert(val.first);
+            if (trustedKeys_.erase(val.first) == 0)
+                trustChanges.added.insert(calcNodeID(val.first));
         }
 
         for (auto const& k : trustedKeys_)
             trustChanges.removed.insert(calcNodeID(k));
+
         trustedKeys_ = std::move(newTrustedKeys);
     }
 
```

Comment:
My hunch tells me that there is a more efficient way to do this for the normal case. The normal case is where there are no changes to the validator lists. Namely:

for each key in trustedKeys_:
  if key not in keyListings_ or key is revoked:
    trustedKeys_.erase(key)
    setUntrusted(key)

for each key in keyListings_:
  res = emplace key in trustedKeys_
  if res was already there:
    setUntrusted(key)

In that case, each list is scanned once for presence of each item in the other list. Also the revocation list is scanned for each item. No changes to any objects are made.

I believe that this would be more efficient than creating a new set, destructing the old, and then inserting and erasing each item into the new and old objects, respectively.

And in the event of a change, such as either through expiration or new/removed entries, just doing the setTrusted() or
setUntrusted() immediately is better than creating a struct of 2 sets to keep track of the changes, to then walk through to feed to setTrusted() and setUntrusted().

If it's important to log the number of changes, then simple counters can be created and incremented.

And even in a case where there are changes to be made to the validator list, I'm not sure how much would be gained, if anything, performance-wise doing things the way implemented currently in this PR.

Reply:
The list of removed trusted validators still needs to be tracked so that it can be fed to Consensus::startRound() to erase recent peer positions. Then again, why not wipe those out in real time too?

Reply:
I believe the change in UNL should be seen atomically by the `Validations` code called by `Consensus::startRound`. This not only changes a validations trust/untrusted status, but also determines what contributes to the preferred ledger calculation.  Since `Validations` members can be called by many parts of the code, I think it is important that its internal state always reflects a complete UNL, rather than a UNL in transition.

Reply:
I've added https://github.com/ripple/rippled/pull/2613/commits/b66aa3810f50150e0dc13afa36b74f7237bc5001 with your suggestion to not re-populate trustedKeys_ every time. One difference from your pseudocode was that we have to check if the key is revoked in both loops.
I'll look into an alternative to the `TrustedChanges` struct that meets @bachase 's atomicity requirement, but I'd prefer such a change (involving Validations code) to be in a separate PR.

Reply:
Good catch--though I'm not sure if the extra check to revocation check tips the scales. This is a micro-optimization to be sure, in either direction. Agreed that any changes to TrustedChanges should be a separate task, and it's unlikely to even be necessary or worth the effot.